src/validation/schema.js

```diff
@@ -322,13 +378,203 @@ const validateEntityFields = (defs, def) =>
+            `Import must be one of "Name" or { name: "Name", as: "Alias" }`,
+        }),
+      )
+    }
+    return type.fields.reduce((errors, field) => {
+      if (!['name', 'as'].includes(field.name.value)) {
+        return errors.push(
+          immutable.fromJS({
+            loc: directive.name.loc,
+            entity: def.name.value,
+            message: `@import field '${field.name.value}' invalid, may only be one of: [name, as]`,
+          }),
+        )
+      }
+      if (field.value.kind != 'StringValue') {
+        return errors.push(
+          immutable.fromJS({
+            loc: directive.name.loc,
+            entity: def.name.value,
+            message: `@import fields [name, as] must be strings`,
+          }),
+        )
+      }
+      return errors
+    }, errors)
+  },
+}
+
+const validateImportDirectiveType = (def, directive, type) => {
+  return importDirectiveTypeValidators[type.kind]
+    ? importDirectiveTypeValidators[type.kind](def, directive, type)
+    : List().push(
+        immutable.fromJS({
+          loc: directive.name.loc,
+          entity: def.name.value,
+          message:
+            `Import must be one of "Name" or { name: "Name", as: "Alias" }`,
+        }),
+      )
+}
+
+const validateImportDirectiveArgumentTypes = (def, directive, argument) => {
+  if (argument.value.kind != 'ListValue') {
+    return List().push(
+      immutable.fromJS({
+        loc: directive.name.loc,
+        entity: def.name.value,
+        message: `@import argument: 'types' must be an list`,
+      }),
+    )
+  }
+
+  return argument.value.values.reduce(
+    (errors, type) => errors.concat(validateImportDirectiveType(def, directive, type)),
+    List(),
+  )
+}
+
+const validateImportDirectiveArgumentFrom = (def, directive, argument) => {
+  if (argument.value.kind != 'ObjectValue') {
+    return List().push(
+      immutable.fromJS({
+        loc: directive.name.loc,
+        entity: def.name.value,
+        message: `@import directive argument: 'from' must be an object`,
+      }),
+    )
+  }
+
+  if (argument.value.fields.length != 1) {
+    return List().push(
+      immutable.fromJS({
+        loc: directive.name.loc,
+        entity: def.name.value,
+        message: `@import argument: 'from' must have an 'id' or 'name' field`,
+      }),
+    )
+  }
+
+  return argument.value.fields.reduce((errors, field) => {
+    if (!['name', 'id'].includes(field.name.value)) {
+      return errors.push(
+        immutable.fromJS({
+          loc: field.name.loc,
+          entity: def.name.value,
+          message: `@import argument: 'from' must be one of { name: "Name" } or { id: "ID" }`,
+        }),
+      )
+    }
+    if (field.value.kind != 'StringValue') {
+      return errors.push(
+        immutable.fromJS({
+          loc: field.name.loc,
+          entity: def.name.value,
+          message: `@import argument: 'from' must be one of { name: "Name" } or { id: "ID" } with string values`,
```

Comment:
I'd drop the `:` after all these messages. Like below where we have `@import argument 'types' must be specified`, I think these can all be `@import argument 'from' must be one of...`. The read flow is better that way, IMHO.